File: apps/web/lib/types/material-purchases.ts
// ── List ──────────────────────────────────────────────────────────────────────

export interface PurchaseListItem {
  id: string
  project_id: string
  project_name: string
  buyer_id: string
  buyer_name: string
  purchased_at: string
  items_count: number
  receipt_file_url: string | null
  receipt_original_filename: string | null
  notes: string | null
  created_at: string
}

export interface PurchasePagination {
  page: number
  per_page: number
  total: number
  total_pages: number
}

export interface PurchaseListResponse {
  data: PurchaseListItem[]
  pagination: PurchasePagination
}

// ── Detail ────────────────────────────────────────────────────────────────────

export interface PurchaseItemDetail {
  id: string
  project_material_id: string
  material_name: string
  material_code: string | null
  quantity: number
  unit: string
}

export interface PurchaseDetail {
  id: string
  project: { id: string; name: string; address?: string }
  buyer: { id: string; full_name: string }
  purchased_at: string
  receipt_file_url: string | null
  receipt_original_filename: string | null
  notes: string | null
  items: PurchaseItemDetail[]
  created_at: string
}

// ── Form data ─────────────────────────────────────────────────────────────────

export interface PurchaseFormProject {
  id: string
  name: string
}

export interface PurchaseFormMaterial {
  id: string
  material_name: string
  material_code: string | null
  available_quantity: number
  unit: string
}

export interface PurchaseFormDataResponse {
  projects: PurchaseFormProject[]
  materials: PurchaseFormMaterial[]
}

// ── Local form state ──────────────────────────────────────────────────────────

export interface PurchaseItemDraft {
  _tempId: string
  project_material_id: string
  material_name: string
  material_code: string | null
  quantity: number
  unit: string
}

// ── Filter ────────────────────────────────────────────────────────────────────

export interface PurchaseFilter {
  project_id?: string
  buyer_id?: string
  date_from?: string
  date_to?: string
  search?: string
  page?: number
  per_page?: number
}

File: services/api/dto/material_purchases.go
package dto

import "time"

// ── Create request ─────────────────────────────────────────────────────────────

type CreatePurchaseItemRequest struct {
	ProjectMaterialID string  `json:"project_material_id"`
	Quantity          float64 `json:"quantity"`
	Unit              string  `json:"unit"`
}

type CreatePurchaseRequest struct {
	ProjectID   string
	PurchasedAt time.Time
	Notes       *string
	Items       []CreatePurchaseItemRequest
}

// ── List ──────────────────────────────────────────────────────────────────────

type PurchaseListItem struct {
	ID                      string    `json:"id"`
	ProjectID               string    `json:"project_id"`
	ProjectName             string    `json:"project_name"`
	BuyerID                 string    `json:"buyer_id"`
	BuyerName               string    `json:"buyer_name"`
	PurchasedAt             time.Time `json:"purchased_at"`
	ItemsCount              int       `json:"items_count"`
	ReceiptFileURL          *string   `json:"receipt_file_url"`
	ReceiptOriginalFilename *string   `json:"receipt_original_filename"`
	Notes                   *string   `json:"notes"`
	CreatedAt               time.Time `json:"created_at"`
}

type PurchasePagination struct {
	Page       int `json:"page"`
	PerPage    int `json:"per_page"`
	Total      int `json:"total"`
	TotalPages int `json:"total_pages"`
}

type PurchaseListResponse struct {
	Data       []PurchaseListItem `json:"data"`
	Pagination PurchasePagination `json:"pagination"`
}

// ── Detail ────────────────────────────────────────────────────────────────────

type PurchaseItemDetail struct {
	ID                string  `json:"id"`
	ProjectMaterialID string  `json:"project_material_id"`
	MaterialName      string  `json:"material_name"`
	MaterialCode      *string `json:"material_code"`
	Quantity          float64 `json:"quantity"`
	Unit              string  `json:"unit"`
}

type PurchaseProjectRef struct {
	ID      string  `json:"id"`
	Name    string  `json:"name"`
	Address *string `json:"address,omitempty"`
}

type PurchaseBuyerRef struct {
	ID       string `json:"id"`
	FullName string `json:"full_name"`
}

type PurchaseDetail struct {
	ID                      string               `json:"id"`
	Project                 PurchaseProjectRef   `json:"project"`
	Buyer                   PurchaseBuyerRef     `json:"buyer"`
	PurchasedAt             time.Time            `json:"purchased_at"`
	ReceiptFileURL          *string              `json:"receipt_file_url"`
	ReceiptOriginalFilename *string              `json:"receipt_original_filename"`
	Notes                   *string              `json:"notes"`
	Items                   []PurchaseItemDetail `json:"items"`
	CreatedAt               time.Time            `json:"created_at"`
}

// ── Form data ─────────────────────────────────────────────────────────────────

type PurchaseFormProject struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type PurchaseFormMaterial struct {
	ID                string  `json:"id"`
	MaterialName      string  `json:"material_name"`
	MaterialCode      *string `json:"material_code"`
	AvailableQuantity float64 `json:"available_quantity"`
	Unit              string  `json:"unit"`
}

type PurchaseFormDataResponse struct {
	Projects  []PurchaseFormProject  `json:"projects"`
	Materials []PurchaseFormMaterial `json:"materials"`
}

// ── Filter ────────────────────────────────────────────────────────────────────

type PurchaseFilter struct {
	ProjectID *string
	BuyerID   *string
	DateFrom  *string
	DateTo    *string
	Search    *string
}
